### PR TITLE
Clamav App

### PR DESCRIPTION
**Pull request by juckerf**
should fulfill #58
A ClamAv App which communicates with a clamav daemon (means of connection are configurable) to scan uploaded files. the scanning is async. if a virus is found the file is deleted.

### CODE REVIEW

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        try {
+            // Create a new socket instance
+            $socket = (new \Socket\Raw\Factory())->createClient($this->socket);
+        } catch (\Exception $e) {
+            throw new Exception('scan of file [' . $file->getId() . '] failed: ' . $e->getMessage());
+        }
+
+        try {
+            // Create a new instance of the Client
+            $quahog = new \Xenolope\Quahog\Client($socket, 30, PHP_NORMAL_READ);
```
Comment (raffis) [inline src/app/Balloon.App.ClamAv/src/lib/Cli.php]: declare all classes as "use" statements:
use \Xenolope\Quahog\Client as ClamAV;

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * Scan file
+     *
+     * @param  File $file
+     * @return int
+     */
+    public function scan(File $file): int
+    {
+        if ($file->getSize() > $this->maxStreamSize) {
```
Comment (raffis) [inline src/app/Balloon.App.ClamAv/src/lib/Cli.php]: methods and variables must use underscore, see CONTRIBUTING.md (code policy)

File: config/config.example.xml [hunk 3]
```diff
@@ -42,26 +42,30 @@
...
                 <hosts>
                     <server>http://localhost:9200</server>
                 </hosts>
-                <index>balloon</index> 
+                <index>balloon</index>
             </config>
         </elasticsearch>
 
+        <ClamAv enabled="0" class="\Balloon\App\ClamAv">
+            <config/>
```
Comment (raffis) [inline config/config.example.xml]: example config should declare all possible configurations socket/max_stream_size

**Comment by raffis** [comment]
@juckerf looking good.
Can you add an additional option aggressive_level:
0 - no action
1 - set app attribute for a node found with a threat: quarantine => true
2 - delete node (trash) and also set quarantine => true
3 - force delete node (what you have now) and should be the default

if nothing found also add an app attribute: quarantine => false

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php (hunk 2 above)
Comment (raffis) [reply src/app/Balloon.App.ClamAv/src/lib/Cli.php]: sry of course I meant properties, methods are fine with camelCase

Comment (raffis) [reply src/app/Balloon.App.ClamAv/src/lib/Cli.php]: variables, class properties and config values => underscore

File: config/config.example.xml (hunk 3 above)
Comment (juckerf) [reply config/config.example.xml]: fixed in #70b21c2

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php (hunk 2 above)
Comment (juckerf) [reply src/app/Balloon.App.ClamAv/src/lib/Cli.php]: fixed in #70b21c2

**Comment by juckerf** [comment]
thx for the review
fixed the issues (code style, example config) in commit 70b21c2.
the different levels are implemented in the same commit too.
commit ed4ed7d adds some tests for the app

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        try {
+            // Create a new socket instance
+            $socket = (new SocketFactory())->createClient($this->socket);
+        } catch (\Exception $e) {
+            throw new Exception('scan of file [' . $file->getId() . '] failed: ' . $e->getMessage());
+        }
+
+        try {
+            // Create a new instance of the Client
+            $clamav = new ClamAv($socket, 30, PHP_NORMAL_READ);
```
Comment (raffis) [inline src/app/Balloon.App.ClamAv/src/lib/Cli.php]: timeout should be configurarble via setOptions wheres 30 is the default.

File: src/app/Balloon.App.ClamAv/src/lib/Cli.php [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        foreach($config as $option => $value) {
+            switch($option) {
+                case 'socket':
+                    $this->socket = (string)$value;
+                    break;
+                case 'max_stream_size':
+                    $this->max_stream_size = (int)$value;
+                    break;
+                case 'aggressiveness':
+                    if ((int)$value <= 3) {
```
Comment (raffis) [inline src/app/Balloon.App.ClamAv/src/lib/Cli.php]: If you check an option here you should throw an Exception (Balloon\App\ClamAv\Exception) if the value is not correct

**Comment by juckerf** [comment]
4e75a5a fixes both "issues" from the last review

**Pull request merged**